File: scripts/pagespeed.py
#!/usr/bin/env python3
"""
PageSpeed Insights API Client

完全免费，只需 Google Cloud API Key（无需 OAuth）。
获取 Core Web Vitals 字段数据（真实用户）+ Lighthouse 实验室数据。

使用方法:
    python scripts/pagespeed.py https://example.com YOUR_API_KEY

API Key 获取: https://developers.google.com/speed/docs/insights/v5/get-started
"""

import sys
import json
import requests

ENDPOINT = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed"


def run_pagespeed(url: str, api_key: str = "", strategy: str = "mobile") -> dict:
    """调用 PageSpeed Insights API，返回结构化的 CWV + Lighthouse 数据。"""
    params = {
        "url": url,
        "strategy": strategy,
        "category": "PERFORMANCE,SEO,ACCESSIBILITY,BEST_PRACTICES",
    }
    if api_key:
        params["key"] = api_key

    resp = requests.get(ENDPOINT, params=params, timeout=30)
    resp.raise_for_status()
    return parse_result(resp.json())


def parse_result(data: dict) -> dict:
    """解析 PSI API 返回的原始 JSON。"""
    result = {
        "url": data.get("id"),
        "strategy": data.get("analysisUTCTimestamp", ""),
    }

    # 字段数据：Chrome User Experience Report（真实用户）
    metrics = data.get("loadingExperience", {}).get("metrics", {})
    result["fieldData"] = {
        "LCP": extract_metric(metrics.get("LARGEST_CONTENTFUL_PAINT_MS")),
        "INP": extract_metric(metrics.get("INTERACTION_TO_NEXT_PAINT")),
        "CLS": extract_metric(metrics.get("CUMULATIVE_LAYOUT_SHIFT_SCORE")),
        "FCP": extract_metric(metrics.get("FIRST_CONTENTFUL_PAINT_MS")),
        "TTFB": extract_metric(metrics.get("EXPERIMENTAL_TIME_TO_FIRST_BYTE")),
        "overallCategory": data.get("loadingExperience", {}).get("overall_category"),
    }

    # 实验室数据：Lighthouse
    lh = data.get("lighthouseResult")
    if lh:
        cats = lh.get("categories", {})
        result["labData"] = {
            "performanceScore": to_100(cats.get("performance", {}).get("score")),
            "seoScore": to_100(cats.get("seo", {}).get("score")),
            "accessibilityScore": to_100(cats.get("accessibility", {}).get("score")),
            "bestPracticesScore": to_100(cats.get("best-practices", {}).get("score")),
            "audits": extract_audits(lh.get("audits", {})),
        }

    return result


def extract_metric(m):
    if not m:
        return None
    return {
        "value": m.get("percentile") if m.get("percentile") is not None else m.get("numericValue"),
        "unit": m.get("numericUnit"),
        "category": m.get("category"),
    }


def to_100(score):
    return round(score * 100) if score is not None else None


def extract_audits(audits: dict) -> dict:
    keys = [
        "largest-contentful-paint",
        "total-blocking-time",
        "cumulative-layout-shift",
        "server-response-time",
        "redirects",
        "unused-javascript",
        "unused-css-rules",
        "modern-image-formats",
        "render-blocking-resources",
        "uses-long-cache-ttl",
    ]
    extracted = {}
    for key in keys:
        a = audits.get(key)
        if a:
            extracted[key] = {
                "score": a.get("score"),
                "displayValue": a.get("displayValue"),
                "title": a.get("title"),
            }
    return extracted


def main():
    if len(sys.argv) < 2:
        print("Usage: python scripts/pagespeed.py <URL> [API_KEY] [mobile|desktop]")
        sys.exit(1)

    url = sys.argv[1]
    api_key = sys.argv[2] if len(sys.argv) > 2 else ""
    strategy = sys.argv[3] if len(sys.argv) > 3 else "mobile"

    try:
        data = run_pagespeed(url, api_key, strategy)
        print(json.dumps(data, indent=2, ensure_ascii=False))
    except requests.HTTPError as e:
        print(f"API Error: {e.response.status_code} - {e.response.text}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/pagespeed.js
#!/usr/bin/env node
/**
 * PageSpeed Insights API Client — 零依赖 Node.js 版本
 *
 * 完全免费，只需 Google Cloud API Key（无需 OAuth）。
 * 获取 Core Web Vitals 字段数据（真实用户）+ Lighthouse 实验室数据。
 *
 * 使用方法:
 *   node scripts/pagespeed.js https://example.com YOUR_API_KEY
 *
 * API Key 获取: https://developers.google.com/speed/docs/insights/v5/get-started
 */

const ENDPOINT = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed";

async function runPageSpeed(url, apiKey, strategy = "mobile") {
  const params = new URLSearchParams({
    url,
    strategy,
    category: "PERFORMANCE,SEO,ACCESSIBILITY,BEST_PRACTICES",
  });
  if (apiKey) params.set("key", apiKey);

  const res = await fetch(`${ENDPOINT}?${params}`, { method: "GET" });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(`PSI API ${res.status}: ${err.error?.message || res.statusText}`);
  }
  return parseResult(await res.json());
}

function parseResult(data) {
  const result = { url: data.id };

  // 字段数据：Chrome User Experience Report（真实用户）
  const metrics = data.loadingExperience?.metrics || {};
  result.fieldData = {
    LCP: extractMetric(metrics.LARGEST_CONTENTFUL_PAINT_MS),
    INP: extractMetric(metrics.INTERACTION_TO_NEXT_PAINT),
    CLS: extractMetric(metrics.CUMULATIVE_LAYOUT_SHIFT_SCORE),
    FCP: extractMetric(metrics.FIRST_CONTENTFUL_PAINT_MS),
    TTFB: extractMetric(metrics.EXPERIMENTAL_TIME_TO_FIRST_BYTE),
    overallCategory: data.loadingExperience?.overall_category,
  };

  // 实验室数据：Lighthouse
  const lh = data.lighthouseResult;
  if (lh) {
    const cats = lh.categories || {};
    result.labData = {
      performanceScore: to100(cats.performance?.score),
      seoScore: to100(cats.seo?.score),
      accessibilityScore: to100(cats.accessibility?.score),
      bestPracticesScore: to100(cats["best-practices"]?.score),
      audits: extractAudits(lh.audits || {}),
    };
  }

  return result;
}

function extractMetric(m) {
  if (!m) return null;
  return {
    value: m.percentile ?? m.numericValue,
    unit: m.numericUnit,
    category: m.category,
  };
}

function to100(score) {
  return score != null ? Math.round(score * 100) : null;
}

function extractAudits(audits) {
  const keys = [
    "largest-contentful-paint",
    "total-blocking-time",
    "cumulative-layout-shift",
    "server-response-time",
    "redirects",
    "unused-javascript",
    "unused-css-rules",
    "modern-image-formats",
    "render-blocking-resources",
    "uses-long-cache-ttl",
  ];
  const extracted = {};
  for (const key of keys) {
    const a = audits[key];
    if (a) {
      extracted[key] = {
        score: a.score,
        displayValue: a.displayValue,
        title: a.title,
      };
    }
  }
  return extracted;
}

// CLI
const [,, url, apiKey, strategy] = process.argv;
if (!url) {
  console.error("Usage: node scripts/pagespeed.js <URL> [API_KEY] [mobile|desktop]");
  process.exit(1);
}

runPageSpeed(url, apiKey || "", strategy || "mobile")
  .then(d => console.log(JSON.stringify(d, null, 2)))
  .catch(e => {
    console.error("Error:", e.message);
    process.exit(1);
  });
